### PR TITLE
Add content for the "Explaining the Internet" session

### PR DESCRIPTION
**Pull request by teodutu**
Signed-off-by: Teodor Dutu <teodor.dutu@gmail.com>

### CODE REVIEW

**Comment by teodutu** [comment]
Some of the images might be proprietary. I will have to double check their licenses or change them.

File: explaining-the-internet/README.md [hunk 1]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
 
 ## Introduction
 
-Objectives and rationale for the current session.
+In this session, we'll discuss some networking fundamentals.
+By doing so, we'll lay the foundations for some of our later sessions regarding networking and the Interned, such as [(Re)Discovering the Browser](./rediscovering-the-browser)
```
Comment (Adoie) [inline explaining-the-internet/README.md:10]: Internet*

File: explaining-the-internet/README.md [hunk 2]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
 
 ## Reminders and Prerequisites
 
-- Information required for this section
-- Commands / snippets that should be known, useful to copy-paste throughout the practical session
+For this session, you'll need:
+- a working internet connection;
+- a Linux environment that you're comfortable working with;
+- to understand what a *process* is;
+- general knowledge of the Linux CLI, as discussed in session [Welcone to Linux](../welcome-to-linux).
```
Comment (Adoie) [inline explaining-the-internet/README.md:22]: Welcome*

File: explaining-the-internet/README.md [hunk 3]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+So what happens here is that the computer on the left is trying to **send** data to the one on the right.
+We call the left computer **the sender** and the right one **the receiver**.
+The two computers are connected via a **medium** on which the actual bits of data that make up the communication are sent.
+In short, this medium is the connection between the aforementioned computers.
+
+However, there are 4 boxes above the aforementioned medium.
+These boxes are altogether known as **The TCP/IP Network Stack**.
+We'll see what's with the *TCP/IP* part soon enough.
+Individually, each box is a **layer**.
+Here are the 4 layers of the The TCP/IP stack:
```
Comment (Adoie) [inline explaining-the-internet/README.md:44]: the*

File: explaining-the-internet/README.md [hunk 4]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+This time called **Round Trip Time (RTT)**.
+As we said, `traceroute` sends and monitors 3 packets in order to display the consistency of the link to each hop.
+
+The first hop is between the VM and the host.
+The other is the router to which the host is connected.
+From then on, it's whe wild Internet itself.
+
+### IPs
+
+IPs are 32-bit (i.e. 4 bytes) numbers used to identify **hosts**, hence they are often called _ip addresses_.
```
Comment (Adoie) [inline explaining-the-internet/README.md:183]: Aici cred ca ar putea fi explicat si de la ce vine IP (Internet Protocol), pentru mai multa claritate. Am observat ca este mentionat la summary, dar nu cred ca ar strica si aici)

File: explaining-the-internet/README.md [hunk 5]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+### Routers
+
+A router is a networking device that connects two or more networks.
+Think of the router you most probably have at home.
+There is a *local network*, to which you can connect via WiFi, or Ethernet cables.
+You're probably doing this right now.
+Another network is the one provided by your Internet Service Provider (ISP).
+
+There are many ISPs around the world and even more local home networks.
+The web thus becomes a _web of rounters_ who transmit **packets** from one network to another.
```
Comment (Adoie) [inline explaining-the-internet/README.md:139]: routers*

File: explaining-the-internet/README.md [hunk 6]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+- **wired**: using a cable on which digital signals that encode the bits are sent;
+- **wireless**: using high-frequency radio waves in order to send the same signals as before
+
+## The Link Layer
+
+This layer is also known as the **Data Link** layer.
+
+It represents the underlying technology of any application.
+The device on which the application is running may have a choice of many technologies to connect to a network, such as Ethernet, WiFi, Bluetooth, 4G, 5G etc.
+Where more than one link exists, the operating system of the device chooses the most approprite link, e.g. a mobile phone might be connected to both 4G and WiFi.
```
Comment (Adoie) [inline explaining-the-internet/README.md:104]: appropriate

File: explaining-the-internet/README.md [hunk 7]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+Data is passed from a process on the sender to a process on the receiver by using the services of the layer below.
+It is only the final layer that actually causes the data to be transmitted onto the transmission media (e.g. cable). Below is a representation of this whole mechanism.
+
+![General Workings of a Network Stack Layer](./assets/network_layers.png)
+
+This method of modular communication is called **encapsulation**.
+It can be likened to taking the data from the layer above and placing it in an **envelope**, writing some additional information on the envelope, then passing the envelope to the layer below.
+At the receiving end, a layer is passed an envelope from the layer below.
+It looks at the data on the outside of the envelope to decide what to do with the contents of the envelope.
+It then open the envelope and pass the contained data up to the next level. The reverse of encapsulation is **decapsulation**.
```
Comment (Adoie) [inline explaining-the-internet/README.md:74]: opens*
passes*

File: explaining-the-internet/README.md [hunk 8]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+These interfaces can be virtual or physical (present on your NIC).
+In your Kali VM, all interfaces are virtual.
+
+**`lo`** is the **loopback** interface.
+It's an internal interface that each host possesses.
+Its purpose is for testing the OS's TCP/IP stack.
+Its IP is, by convention `127.0.0.1`, or `0.0.0.0`.
+
+**`eth0`** is the interface used to connect to the Internet.
+It acts as a middle man between your host OS (the one in which you're running the VM) and the guest OS (Kali).
```
Comment (Adoie) [inline explaining-the-internet/README.md:225]: middleman*

File: explaining-the-internet/README.md [hunk 9]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+In the next section, we'll explain all of these components.
+
+#### Anatomy of a URL
+
+The image below summarises the building blocks of a URL
+
+![Anatomy of a URL](./assets/url_anatomy.png)
+
+Let's disect these components:
+- **The scheme** (sometimes called _schema_) indicates the application layer protocol that the browser must use to request the resouce.
```
Comment (Adoie) [inline explaining-the-internet/README.md:301]: resource*

File: explaining-the-internet/README.md [hunk 10]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+- **The anchor**, also known as **fragment**, is like a _bookmark_ to some specific part of the resource.
+It gives the browser directions to show the content located at that _bookmarked_ spot.
+On an HTML document, for example, the browser will scroll to the point where the anchor is defined; on a video or audio document, the browser will try to go to the time the anchor represents.
+Markdown documents also use anchors, like so: https://github.com/security-summer-school/essentials/tree/master/explaining-the-internet#anatomy-of-a-url.
+
+You might have heard the same things being called _Uniform Resource Identifiers (URIs)_.
+This is correct, but it's not the most precise name you can use.
+In order to understand the difference between URLs and URIs, check out their [corresponding section](#url-vs-uri).
+
+We'll se how URLs are translated into IPs by a naming system called the **Domanin Name System (DNS)**.
```
Comment (Adoie) [inline explaining-the-internet/README.md:328]: see*

File: explaining-the-internet/README.md [hunk 11]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+
+Remember: IPs, **not URLs**, are what's used to identify hosts (web servers included) in the Internet.
+So there needs to be a service by which the URLs we've just learned about are mapped to IPs.
+This service is called the DNS.
+It organises URLs in a hierarchical manner, in order to rapidly respond to queries.
+You can think of the DNS as being similar to a **phonebook of the Internet**.
+
+Here's how all of this works.
+When you type `security-summer-school.github.io` into your browser, it sends a query over the Internet in order to find the website associated with the URL `security-summer-school.github.io`.
+A query is a question asking to look up the domain name and respond the corresponging IP address.
```
Comment (Adoie) [inline explaining-the-internet/README.md:340]: corresponding*

File: explaining-the-internet/README.md [hunk 12]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+Let's follow the path of this question accross the Internet.
```
Comment (Adoie) [inline explaining-the-internet/README.md:341]: across*

File: explaining-the-internet/README.md [hunk 13]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+
+## The Transport Layer
+
+So with IPs, we can connect two hosts however far apart they are.
+So far, we managed to directly connect hosts.
+Then we moved on to LANs and then to the whole Web.
+This must be it, right?
+We're now worldwide.
+Why all the fuss about the Transport Layer.
+Let's run an experiment: in your browser, open to tabs to your favourite website (which is https://security-summer-school.github.io, of course).
```
Comment (Adoie) [inline explaining-the-internet/README.md:408]: two tabs*

File: explaining-the-internet/README.md [hunk 14]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+How are those connections handled?
+How come what you do in one of the browsers does not interfere with what you do in the other?
```
Comment (Adoie) [inline explaining-the-internet/README.md:410]: one of the tabs*

File: explaining-the-internet/README.md [hunk 15]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+Upon receiving a NACK, the sender **retransmits** the segment.
+The same mechanism is employed for segments that do not make it to the receiver.
+
+In short, TCP makes sure that whatever is sent is received completely and correctly.
+This feature is obviously important.
+As a result, TCP is used for the transport-layer in the following application-level protocols and scenarios:
+- **SMTP (Simple Mail Transfer Protocol)**, **IMAP (Internet Access Message Protocol)** and **POP3 (Post Office Protocol version 3)**: the main protocols used by email services worldwide.
+- **HTTP/HTTPS (HyperText Transfer Protocol / Secured)**: the most popular protocol used to transfer web pages.
+By default, HTTP uses port 80 and HTTPS uses port 443.
+- **SSH (Secure Shell)**: a wildey used protocol for obtaining a remote shell on another machine.
```
Comment (Adoie) [inline explaining-the-internet/README.md:473]: widely*

File: explaining-the-internet/README.md [hunk 16]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+
+#### SSH
+
+We've been using `ssh` to connect to the remote machines used for challenges and tutorials.
+Now we know it is a protocol used for connecting to remote hosts via a shell.
+It opens a TCP connection to the remote host, where each character typed in the terminal is sent to the receiving host.
+
+As its name suggests, the main gist of this protocol is **secured**.
```
Comment (Adoie) [inline explaining-the-internet/README.md:481]: "main gist" cred ca e pleonasm, sau, cel putin, e suficient "gist"

File: explaining-the-internet/README.md [hunk 17]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+The security implies that the characters are not sent to the receiver as plain text, but are **encrypted**.
+While we'll discuss more about encryption in the [Data Security](../data-security) session, it is important that we outline the the main workings of encryption.
```
Comment (Adoie) [inline explaining-the-internet/README.md:483]: un singur "the"*

File: explaining-the-internet/README.md [hunk 18]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+By default, SSH uses port 22. We can specify another port by using the `-p` parameter like so:
+```
+$ ssh user@ip -p port
+```
+
+### UDP
+
+We said TCP guarantees the correctness of the delivered data, which is great.
+This guarantee, however, comes at the cost of performance.
+In the pictura above, we saw how ACKs are transmitted.
```
Comment (Adoie) [inline explaining-the-internet/README.md:508]: picture*

File: explaining-the-internet/README.md [hunk 19]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+- **UDP:** User Datagrem Protocol
+- **URI:** Uniform Resource Identifier
+- **URL:** Uniform Resource Locator
+
+Here's a short summary of the layers of the TCP/IP stack:
+- **Link:** Provides us with direct connections to other hosts.
+Also adds identifiers to NICs in the form of MAC addresses.
+- **Internet:** Provides routing and identifiers for hosts in the Internet, in the form of IP addresses.
+- **Transport:** Provides connections between processes on different hosts by using ports.
+- **Application:** Composes actual the message to be delivered to the receiver.
```
Comment (Adoie) [inline explaining-the-internet/README.md:588]: the actual message*

File: explaining-the-internet/README.md [hunk 20]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+In the session, we discussed the TCP/IP stack, which is implemented by the kernel of every Operating System (Linux included).
+In the early days of the Internet, however, another model was popular.
+This was the **OSI Model**.
+As a matter of fact, it's still a popular model for teaching about the internet, because it makes a more rigorous separation of what we called the Application Layer.
+This layer is split into 3, and the transmission medium is added as a separate physical layer, thus resulting a 7-layer stack.
+You can learn about this alternative stack [here](https://www.imperva.com/learn/application-security/osi-model/), but keep in mind that it is almost never used in practice.
+
+### The Networking Bible
+
+Speaking of the OSI model, it is paramounnt that we mention the most influential book in computer networking to this day, namely Andrew Tanenbaum's [Computer Networks](http://index-of.es/Varios-2/Computer%20Networks%205th%20Edition.pdf).
```
Comment (Adoie) [inline explaining-the-internet/README.md:634]: paramount*

File: explaining-the-internet/README.md [hunk 21]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+Dating from 1981 (about the time these stacks were first defined), this book has been the cornerstone of networking ever since. It's a must-read for every netowrking enthusiast.
```
Comment (Adoie) [inline explaining-the-internet/README.md:635]: networking*

File: explaining-the-internet/README.md [hunk 22]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+[Here](https://danielmiessler.com/study/difference-between-uri-url/)'s the difference explained.
+
+### IPv6
+
+Since regular IPs (also known as **IPv4**) are 32-bit numbers, there are `2^32` possible unique IPs.
+This number may look large, but it's _just_ 4,294,967,296, so just upwards of 4 billion.
+But the world's population is 7.9 billion as of September 2021, and growing.
+The pigeonhole principle tells us there are not enough IPs so that every human can be connected to the Internet at the same time.
+
+For this reason and some others relatd to the lack of native security of the IP protocol, a new iteration of IPs was designed: **IPv6**.
```
Comment (Adoie) [inline explaining-the-internet/README.md:660]: related*

File: explaining-the-internet/README.md [hunk 23]
```diff
@@ -6,29 +6,666 @@ Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
...
+`2^128` is 39 **digits** long, so it should suffice for the entirety of Earth's population (a few times over).
+
+For a better (but still short) introduction to IPv6, check out [this](http://www.steves-internet-guide.com/ipv6-guide/) source.
+
+### DHCP
+
+**Dynamic Host Configuration Protocol (DHCP)** is a protocol used in order to manage and distribute IP addresses.
+Hosts are *leased* IP addresses from DHCP servers for some limited amounts of time.
+When an address is about to expire, a host can ask the server to extend the lease.
+You can read more about DHCPP servers [here](https://www.infoblox.com/glossary/dhcp-server/).
```
Comment (Adoie) [inline explaining-the-internet/README.md:671]: DHCP*

**Review by Adoie — CHANGES_REQUESTED** [review]

**Review by Adoie — APPROVED** [review]

**Pull request merged**